lib/puppet/provider/ec2_volume/v2.rb

```diff
@@ -46,74 +56,134 @@ def self.volume_to_hash(region, volume)
       volume_type: volume.volume_type,
       availability_zone: volume.availability_zone,
       snapshot_id: volume.snapshot_id,
-      ensure: :present,
-      region: region,
+      ensure: volume_ensure(volume),
+      state: volume.state,
+      region: region
     }
     config[:attach] = attachments unless attachments.empty?
     config
   end
 
+  def self.volume_ensure(volume)
+    if volume.state == 'available'
+      :absent
+    else
+      :present
+    end
+  end
+
   def exists?
     Puppet.info("Checking if EC2 volume #{name} exists in region #{target_region}")
     @property_hash[:ensure] == :present
   end
 
   def create_from_snapshot(config)
-    snapshot = resource[:snapshot_id] ? resource[:snapshot_id] : false
+    snapshot = latest_snapshot
     config['snapshot_id'] = snapshot if snapshot
     config
   end
 
+  def latest_snapshot
+    return resource[:snapshot_id] if resource[:snapshot_id]
+    return false unless resource[:snapshot_label]
+    Puppet.notice("Attempting to fetch snapshot ID")
```

Comment:
Maybe include the label name in the notice text so we know what it's looking for?